pandas/core/indexes/datetimes.py

```diff
@@ -416,6 +416,16 @@ def _formatter_func(self):
     # --------------------------------------------------------------------
     # Set Operation Methods
 
+    def _can_range_setop(self, other) -> bool:
+        # GH 46702: If self or other have non-UTC tzs, DST transitions prevent
+        # range representation due to no singular step
+        if self.tz is not None and not timezones.is_utc(self.tz):
+            return False
+        other_tz = getattr(other, "tz", None)
+        if other_tz is not None and not timezones.is_utc(other_tz):
```

Comment:
fine if you want to do quick fix, but could also include fixed-offset tzs.  also i think we're assured matching types/dtypes at this point

Reply:
Added a check for `not timezones.is_fixed_offset`.

Yeah I think a more complete check would be to add a `.fold` attribute and then check `self.fold.nunique > 1`. That would allow some `DatetimeIndex`es to still use the distpatch to range ops. https://github.com/pandas-dev/pandas/issues/47130